src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java

```diff
@@ -210,68 +211,80 @@ private List<SourceRecord> pollInternal() {
-                    publishFullDocumentOnlyTombstoneOnDelete
-                        && !changeStreamDocument.containsKey(FULL_DOCUMENT);
-                if (publishFullDocumentOnly) {
-                  if (changeStreamDocument.containsKey(FULL_DOCUMENT)
-                      && changeStreamDocument.get(FULL_DOCUMENT).isDocument()) {
-                    valueDocument = Optional.of(changeStreamDocument.getDocument(FULL_DOCUMENT));
-                  }
-                } else {
-                  valueDocument = Optional.of(changeStreamDocument);
-                }
-
-                if (valueDocument.isPresent() || isTombstoneEvent) {
-                  BsonDocument valueDoc = valueDocument.orElse(new BsonDocument());
-                  LOGGER.trace("Adding {} to {}: {}", valueDoc, topicName, sourceOffset);
-
-                  if (valueDoc instanceof RawBsonDocument) {
-                    int sizeBytes = ((RawBsonDocument) valueDoc).getByteBuffer().limit();
-                    statisticsManager.currentStatistics().getMongodbBytesRead().sample(sizeBytes);
-                  }
-
-                  BsonDocument keyDocument;
-                  if (sourceConfig.getKeyOutputFormat() == MongoSourceConfig.OutputFormat.SCHEMA) {
-                    keyDocument = changeStreamDocument;
-                  } else if (sourceConfig.getBoolean(DOCUMENT_KEY_AS_KEY_CONFIG)
-                      && changeStreamDocument.containsKey(DOCUMENT_KEY_FIELD)) {
-                    keyDocument = changeStreamDocument.getDocument(DOCUMENT_KEY_FIELD);
-                  } else {
-                    keyDocument = new BsonDocument(ID_FIELD, changeStreamDocument.get(ID_FIELD));
-                  }
-
-                  createSourceRecord(
-                          keySchemaAndValueProducer,
-                          isTombstoneEvent
-                              ? TOMBSTONE_SCHEMA_AND_VALUE_PRODUCER
-                              : valueSchemaAndValueProducer,
-                          sourceOffset,
-                          topicName,
-                          keyDocument,
-                          valueDoc)
-                      .map(sourceRecords::add);
-                }
-              }
-            });
+      // if isCopying is true, we want to set the COPY_KEY flag so that kafka has context that a
+      // copy is in progress. However, for the last document that we are copying, we should not set
+      // this flag because the copy has completed, otherwise we are relying on future change stream
+      // events to signify that we are no longer copying. We also need to set the _id field to be a
+      // valid resume token, which during copying exists in the cachedResumeToken variable.
+      boolean lastDocument = !batchIterator.hasNext();
+      boolean noMoreDataToCopy = copyDataManager != null && !copyDataManager.isCopying();
+      if (isCopying && lastDocument && noMoreDataToCopy) {
+        sourceOffset.put(ID_FIELD, cachedResumeToken.toJson());
```

Comment:
Instead of `cachedResumeToken `, it might be safer to create and use a different variable here that contains the same value as `cachedResumeToken` but isn't modified anywhere else. `cachedResumeToken` get's set to `null` ([here](https://github.com/mongodb/mongo-kafka/blob/1a09960e1f90c7ab077c1f1d6a502d162f88e5fb/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L697)) within `getResumeToken`, which ends up being called [here](https://github.com/mongodb/mongo-kafka/blob/1a09960e1f90c7ab077c1f1d6a502d162f88e5fb/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L596) if `cursor == null`. The `cursor` could be set to `null` if the cluster runs into a temporary connection issue.

Reply:
Yeah that's a good catch, I'll create a new token called `resumeTokenAfterCopy`

Reply:
Wait I think we're actually good here, I'm happy to create a new variable for extra safety though if you're still concerned.

because we are assigning this in the copied batch the `cursor == null` code won't be hit until after we use and store the `cachedResumeToken`. i.e. the getNextBatch function exits early [here](https://github.com/mongodb/mongo-kafka/blob/1a09960e1f90c7ab077c1f1d6a502d162f88e5fb/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L581).